File: CS/MyGridCtrl/MyBandedGridView.cs
using DevExpress.XtraGrid.Views.BandedGrid;
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraGrid;
using DevExpress.Data;
using DevExpress.XtraGrid.Columns;

namespace DevExpress.S170736 {
    public class MyBandedGridView : BandedGridView {        
        public MyBandedGridView(GridControl ownerGrid)
            : base(ownerGrid) {
        }
        public MyBandedGridView() {
        }
        protected override string ViewName {
            get {
                return "MyBandedGridView";
            }
        }
        public int GetGroupFooterRowCount() {
            if (IsDesignMode) {
                return 0;
            }
            var footerRowCount = GetFooterRowCount();
            if (footerRowCount.Count == 0) {
                return 0;
            }
            return footerRowCount.Max(kvp => kvp.Value);
        }
        private Dictionary<BandedGridColumn, int> GetFooterRowCount() {
            var columnFooterRowCount = new Dictionary<BandedGridColumn, int>();
            foreach (GridGroupSummaryItem summary in GroupSummary) {
                var column = summary.ShowInGroupColumnFooter as BandedGridColumn;
                if (column == null) {
                    continue;
                }
                if (columnFooterRowCount.ContainsKey(column)) {
                    columnFooterRowCount[column]++;
                }                
                if ((!columnFooterRowCount.ContainsKey(column))) {
                    columnFooterRowCount.Add(column, 1);
                }
            }
            return columnFooterRowCount;
        }
        protected override void SynchronizeSummary(System.Collections.IList gridSum, Data.SummaryItemCollection summary) {
            var gridSumCollection = gridSum as GridGroupSummaryItemCollection;
            if (gridSumCollection == null) {
                base.SynchronizeSummary(gridSum, summary);
                return;
            }
            var myGridSumCollection = new List<GridGroupSummaryItem>();
            var groupFooterRowCount = GetGroupFooterRowCount();
            MyGridSumCollectionAddSummary(gridSumCollection, myGridSumCollection);
            var query = GetGroups(myGridSumCollection);
            var tempSummaryItemList = new List<GridGroupSummaryItem>();            
            FillTempSummaryItemList(groupFooterRowCount, query, tempSummaryItemList);
            GridSumCollectionAddSummaries(gridSumCollection, tempSummaryItemList);
            base.SynchronizeSummary(gridSum, summary);
        }
        private void FillTempSummaryItemList(int groupFooterRowCount, IEnumerable<IGrouping<GridColumn, GridGroupSummaryItem>> query, List<GridGroupSummaryItem> tempSummaryItemList) {
            foreach (var group in query) {
                GridGroupSummaryItem summaryItem = null;
                GridColumn showInGroupColumnFooter = null;
                var fieldName = string.Empty;
                var summaryItemList = group.ToList();
                for (var i = 0; i < groupFooterRowCount; i++) {
                    try {
                        summaryItem = summaryItemList[i];
                    } catch {
                        summaryItem = null;
                    }
                    if (summaryItem != null) {
                        fieldName = summaryItem.FieldName;
                        showInGroupColumnFooter = summaryItem.ShowInGroupColumnFooter;
                        continue;
                    }
                    if (fieldName != string.Empty && showInGroupColumnFooter != null) {
                        summaryItem = new GridGroupSummaryItem() { SummaryType = SummaryItemType.None, ShowInGroupColumnFooter = showInGroupColumnFooter, FieldName = fieldName };
                        tempSummaryItemList.Add(summaryItem);
                    }
                }
            }
            AddEmptySummaryColumn(groupFooterRowCount, tempSummaryItemList);
        }
        private void AddEmptySummaryColumn(int groupFooterRowCount, List<GridGroupSummaryItem> tempSummaryItemList) {
            var footerRowCount = GetFooterRowCount();
            foreach (BandedGridColumn col in Columns) {
                if (footerRowCount.ContainsKey(col)) {
                    continue;
                }
                CreateEmptySummaries(col, tempSummaryItemList, groupFooterRowCount);
            }
        }
        private void CreateEmptySummaries(BandedGridColumn col, List<GridGroupSummaryItem> tempSummaryItemList, int groupFooterRowCount) {
            GridGroupSummaryItem summaryItem = null;
            GridColumn showInGroupColumnFooter = null;
            var fieldName = string.Empty;
            for (var i = 0; i < groupFooterRowCount - 1; i++) {
                showInGroupColumnFooter = col;
                fieldName = col.FieldName;
                summaryItem = new GridGroupSummaryItem() { SummaryType = SummaryItemType.None, ShowInGroupColumnFooter = showInGroupColumnFooter, FieldName = fieldName };
                tempSummaryItemList.Add(summaryItem);
            }
        }
        private void GridSumCollectionAddSummaries(GridGroupSummaryItemCollection gridSumCollection, List<GridGroupSummaryItem> tempSummaryItemList) {
            if (tempSummaryItemList.Count > 0) {
                gridSumCollection.AddRange(tempSummaryItemList.ToArray());
            }
        }
        private IEnumerable<IGrouping<GridColumn, GridGroupSummaryItem>> GetGroups(List<GridGroupSummaryItem> gridSumCollection) {
            var groups = from item in gridSumCollection
                        group item by item.ShowInGroupColumnFooter;
            return groups;
        }
        private void MyGridSumCollectionAddSummary(GridGroupSummaryItemCollection gridSumCollection, List<GridGroupSummaryItem> myGridSumCollection) {
            foreach (GridGroupSummaryItem summaryItem in gridSumCollection) {
                myGridSumCollection.Add(summaryItem);
            }
        }
    }
}

File: VB/MyGridCtrl/MyBandedGridView.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraGrid.Views.BandedGrid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraGrid
Imports DevExpress.Data
Imports DevExpress.XtraGrid.Columns

Namespace DevExpress.S170736
    Public Class MyBandedGridView
        Inherits BandedGridView
        Public Sub New(ByVal ownerGrid As GridControl)
            MyBase.New(ownerGrid)
        End Sub
        Public Sub New()
        End Sub
        Protected Overrides ReadOnly Property ViewName() As String
            Get
                Return "MyBandedGridView"
            End Get
        End Property
        Public Function GetGroupFooterRowCount() As Integer
            If IsDesignMode Then
                Return 0
            End If
            Dim footerRowCount = GetFooterRowCount()
            If footerRowCount.Count = 0 Then
                Return 0
            End If
            Return footerRowCount.Max(Function(kvp) kvp.Value)
        End Function
        Private Function GetFooterRowCount() As Dictionary(Of BandedGridColumn, Integer)
            Dim columnFooterRowCount = New Dictionary(Of BandedGridColumn, Integer)()
            For Each summary As GridGroupSummaryItem In GroupSummary
                Dim column = TryCast(summary.ShowInGroupColumnFooter, BandedGridColumn)
                If column Is Nothing Then
                    Continue For
                End If
                If columnFooterRowCount.ContainsKey(column) Then
                    columnFooterRowCount(column) += 1
                End If
                If ((Not columnFooterRowCount.ContainsKey(column))) Then
                    columnFooterRowCount.Add(column, 1)
                End If
            Next summary
            Return columnFooterRowCount
        End Function
        Protected Overrides Function SynchronizeSummary(ByVal gridSum As System.Collections.IList, ByVal summary As Data.SummaryItemCollection) As List(Of SummaryItem)
            Dim gridSumCollection = TryCast(gridSum, GridGroupSummaryItemCollection)
            If gridSumCollection Is Nothing Then
                Return MyBase.SynchronizeSummary(gridSum, summary)
            End If
            Dim myGridSumCollection = New List(Of GridGroupSummaryItem)()
            Dim groupFooterRowCount = GetGroupFooterRowCount()
            MyGridSumCollectionAddSummary(gridSumCollection, myGridSumCollection)
            Dim query = From item In gridSumCollection
                        Let summ = CType(item, GridGroupSummaryItem)
                        Group item By summ.ShowInGroupColumnFooter
                        Into Group

            Dim tempSummaryItemList = New List(Of GridGroupSummaryItem)()
            For Each group In query
                Dim summaryItem As GridGroupSummaryItem = Nothing
                Dim showInGroupColumnFooter As GridColumn = Nothing
                Dim fieldName = String.Empty
                For i = 0 To groupFooterRowCount - 1
                    Try
                        summaryItem = group.Group(i)
                    Catch
                        summaryItem = Nothing
                    End Try
                    If summaryItem IsNot Nothing Then
                        fieldName = summaryItem.FieldName
                        showInGroupColumnFooter = summaryItem.ShowInGroupColumnFooter
                        Continue For
                    End If
                    If fieldName IsNot String.Empty AndAlso showInGroupColumnFooter IsNot Nothing Then
                        summaryItem = New GridGroupSummaryItem() With {.SummaryType = SummaryItemType.None, .ShowInGroupColumnFooter = showInGroupColumnFooter, .FieldName = fieldName}
                        tempSummaryItemList.Add(summaryItem)
                    End If
                Next i
            Next group
            AddEmptySummaryColumn(groupFooterRowCount, tempSummaryItemList)
            GridSumCollectionAddSummaries(gridSumCollection, tempSummaryItemList)
            MyBase.SynchronizeSummary(gridSum, summary)
        End Function
        Private Sub AddEmptySummaryColumn(ByVal groupFooterRowCount As Integer, ByVal tempSummaryItemList As List(Of GridGroupSummaryItem))
            Dim footerRowCount = GetFooterRowCount()
            For Each col As BandedGridColumn In Columns
                If footerRowCount.ContainsKey(col) Then
                    Continue For
                End If
                CreateEmptySummaries(col, tempSummaryItemList, groupFooterRowCount)
            Next col
        End Sub
        Private Sub CreateEmptySummaries(ByVal col As BandedGridColumn, ByVal tempSummaryItemList As List(Of GridGroupSummaryItem), ByVal groupFooterRowCount As Integer)
            Dim summaryItem As GridGroupSummaryItem = Nothing
            Dim showInGroupColumnFooter As GridColumn = Nothing
            Dim fieldName = String.Empty
            For i = 0 To groupFooterRowCount - 2
                showInGroupColumnFooter = col
                fieldName = col.FieldName
                summaryItem = New GridGroupSummaryItem() With {.SummaryType = SummaryItemType.None, .ShowInGroupColumnFooter = showInGroupColumnFooter, .FieldName = fieldName}
                tempSummaryItemList.Add(summaryItem)
            Next i
        End Sub
        Private Sub GridSumCollectionAddSummaries(ByVal gridSumCollection As GridGroupSummaryItemCollection, ByVal tempSummaryItemList As List(Of GridGroupSummaryItem))
            If tempSummaryItemList.Count > 0 Then
                gridSumCollection.AddRange(tempSummaryItemList.ToArray())
            End If
        End Sub
        Private Sub MyGridSumCollectionAddSummary(ByVal gridSumCollection As GridGroupSummaryItemCollection, ByVal myGridSumCollection As List(Of GridGroupSummaryItem))
            For Each summaryItem As GridGroupSummaryItem In gridSumCollection
                myGridSumCollection.Add(summaryItem)
            Next summaryItem
        End Sub
    End Class
End Namespace
